python/pyproject.toml

```diff
@@ -18,6 +18,10 @@
 [build-system]
 requires = [
     "setuptools>=42",
-    "wheel"
+    "wheel",
```

Comment:
You should also add `singledispatch` here which includes the backport of `@singledispatchmethod` for python 3.7 (adding it to tox.ini also should fix the test failure in your latest commit)

Reply:
Thank you @samredai  for your suggestion for py37 `singledispatch` backport, I managed to get it working. But mypy is still unable to redefining the same method and I double checked it's on the [latest 0.931](https://github.com/apache/iceberg/runs/5430406474?check_suite_focus=true#step:5:369) as you suggested. I am scratching my head trying to see what's the correct way to make it understand this is not an error,

Reply:
I think I can temporarily silence the mypy linter error with `# type: ignore[no-redef]` for each method I am overriding. In the meantime I can read more with how mypy works with those

Reply:
I looked a bit closer at what mypy is doing and it's getting confused because you've given actual names to the registered functions. The convention is to only name the generic function and use `_` for the additional type based functions that you're registering (like the example in the [docs](https://docs.python.org/3/library/functools.html#functools.singledispatchmethod)). I confirmed that will pass all mypy inspection.

Reply:
To my comment about adding `singledispatch` and the other dependencies here, I did a little more research and `install_requires` in [setup.cfg](https://github.com/apache/iceberg/blob/master/python/setup.cfg#L40) is actually the right place for this. My bad!
```
[options]
...
install_requires =
    pytz
    python-dateutil
    fastavro>=1.3.2<1.4.0
    python-dateutil
    singledispatch
```